dask_kubernetes/core.py

```diff
@@ -14,19 +17,186 @@
 from distributed.comm.utils import offload
-import kubernetes
+from distributed.utils import Log, Logs
+import kubernetes_asyncio as kubernetes
 from tornado import gen
 
-from .objects import make_pod_from_dict, clean_pod_template
+from .objects import (
+    make_pod_from_dict,
+    make_service_from_dict,
+    clean_pod_template,
+    clean_service_template,
+)
 from .auth import ClusterAuth
-from .logs import Log, Logs
 
 logger = logging.getLogger(__name__)
 
 
-class KubeCluster(Cluster):
+class Pod(ProcessInterface):
+    """ A superclass for Kubernetes Pods
+    See Also
+    --------
+    Worker
+    Scheduler
+    """
+
+    def __init__(self, core_api, pod_template, namespace, loop=None, **kwargs):
+        self.pod = None
+        self.core_api = core_api
+        self.pod_template = copy.deepcopy(pod_template)
+        self.namespace = namespace
+        self.name = None
+        self.loop = loop
+        self.kwargs = kwargs
+        super().__init__()
+
+    @property
+    def cluster_name(self):
+        return self.pod_template.metadata.labels["dask.org/cluster-name"]
+
+    async def start(self, **kwargs):
+        self.pod = await self.core_api.create_namespaced_pod(
+            self.namespace, self.pod_template
+        )
+
+        while self.pod.status.phase == "Pending":
+            self.pod = await self.core_api.read_namespaced_pod(
+                self.pod.metadata.name, self.namespace
+            )
+            await asyncio.sleep(0.1)
+
+        while self.address is None:
+            logs = await self.logs()
+            for line in logs.splitlines():
+                for query_string in ["worker at:", "Scheduler at:"]:
+                    if query_string in line:
+                        self.address = line.split(query_string)[1].strip()
+            await asyncio.sleep(0.1)
+
+        await super().start(**kwargs)
+
+    async def close(self, **kwargs):
+        if self.pod:
+            await self.core_api.delete_namespaced_pod(
+                self.pod.metadata.name, self.namespace
+            )
+        await super().close(**kwargs)
+
+    async def logs(self):
+        return Log(
+            await self.core_api.read_namespaced_pod_log(
+                self.pod.metadata.name, self.namespace
+            )
+        )
+
+    def __repr__(self):
+        return "<Pod %s: status=%s>" % (type(self).__name__, self.status)
+
+
+class Worker(Pod):
+    """ A Remote Dask Worker controled by SSH
+    Parameters
+    ----------
+    scheduler: str
+        The address of the scheduler
+    address: str
+        The hostname where we should run this worker
+    connect_kwargs: dict
+        kwargs to be passed to asyncssh connections
+    kwargs:
+        TODO Document Worker kwargs
+    """
+
+    def __init__(self, scheduler: str, **kwargs):
+        super().__init__(**kwargs)
+
+        self.scheduler = scheduler
+
+        self.pod_template.metadata.labels["dask.org/component"] = "worker"
```

Comment:
We used to have more labels here, didn't we?  I suspect that, for example, Pangeo clusters are expecting certain labels today.

Reply:
Those still exist, common labels are set all together when the pod spec is cleaned. This is just setting a component specific label once we are into differentiated `Scheduler` and `Worker` classes.